crypto/dh/dh_lib.c

```diff
@@ -205,11 +205,16 @@ int DH_set0_pqg(DH *dh, BIGNUM *p, BIGNUM *q, BIGNUM *g)
 
     ffc_params_set0_pqg(&dh->params, p, q, g);
     dh->params.nid = NID_undef;
-    DH_get_nid(dh); /* Check if this is a named group and cache it */
-
-    if (q != NULL)
-        dh->length = BN_num_bits(q);
-
+    /*
+     * Check if this is a named group. If it finds a named group then the
+     * 'q' and 'length' value are either already set or are set by the
+     * call.
+     */
+    if (DH_get_nid(dh) == NID_undef) {
+        /* If its not a named group then set the 'length' if q is not NULL */
+        if (q != NULL)
+            dh->length = BN_num_bits(q);
```

Comment:
This is not new behavior, but the code in 1.1.1 actually ignored the length if q was set, and uses BN_priv_rand_range(q). But from reading the manpage, this wasn't the behavior I expected, I expected it to remain 0. I'm also not sure why it's not using BN_priv_rand_range(p-1). So I think we should either not set dh->length in this case, or update the manpage.

Reply:
Then I think we should correct the manpage.